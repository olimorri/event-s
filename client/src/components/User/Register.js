import React, { useState } from 'react';
import UsersApiService from '../../services/UsersApiService';
import auth from '../../utils/auth';
import {
  Flex,
  Box,
  FormControl,
  FormLabel,
  Input,
  Select,
  Stack,
  Button,
  Heading,
  FormHelperText,
} from '@chakra-ui/react';
import { motion } from 'framer-motion';
const initialState = {
  email: '',
  password: '',
  firstName: '',
  lastName: '',
  photo: '',
  host: '',
};

export default function Register(props) {
  const [state, setState] = useState(initialState);

  const handleChange = (e) => {
    let { name, value } = e.target;
    if (name === 'host') {
      value = value === 'Guest' ? false : true;
    }
    setState((prevState) => ({
      ...prevState,
      [name]: value,
    }));
  };

  const handleSubmit = async (e) => {
    e.preventDefault();

    const { email, password, firstName, lastName, host, photo } = state;
    const user = { email, password, firstName, lastName, host, photo };
    const res = await UsersApiService.register(user);
    if (res.error) {
      alert(`${res.message}`);
      setState(initialState);
    } else {
      props.setIsAuthenticated(true);
      auth.login(() => props.history.push('/profile'));
    }
  };

  const validateForm = () => {
    return (
      !state.email ||
      !state.password ||
      !state.firstName ||
      !state.lastName ||
      !state.host === ''
    );
  };
  return (
    <Flex
      minH={'100vh'}
      align={'center'}
      justify={'center'}
      bg={'gray.100'}
      bgImage="url('https://res.cloudinary.com/dujun1hoe/image/upload/v1615228154/event-s/gradient-background-26046-26731-hd-wallpapers.jpg_cenrqe.png')"
      bgSize="100%"
      backgroundRepeat="no-repeat"
    >
      <Stack spacing={8} w={'50%'} maxW={'lg'} py={12} px={6}>
        <motion.div
          initial={{
            opacity: 0,
            scale: 0,
          }}
          animate={{
            opacity: 1,
            scale: 1,
          }}
          transition={{
            type: 'spring',
            stiffness: 100,
            // @ts-ignore
            ease: 'linear',
          }}
          exit={{
            opacity: 0,
            x: '-100vw',
          }}
        >
          <Stack align={'center'}>
            <Heading color={'white'} fontSize={'4xl'}>
              Create new account
            </Heading>
          </Stack>
          <Box rounded={'lg'} bg={'white'} boxShadow={'lg'} p={8}>
            <Stack spacing={4}>
              <FormControl id="first_name">
                <FormLabel>First name</FormLabel>
                <Input
                  id="first_name"
                  placeholder="First Name"
                  name="firstName"
                  value={state.firstName}
                  onChange={handleChange}
                />
                {/* </FormControl> */}
                {/* <FormControl id="last_name"> */}
                <FormLabel>Last name</FormLabel>
                <Input
                  id="last_name"
                  placeholder="Last Name"
                  name="lastName"
                  value={state.lastName}
                  onChange={handleChange}
                />
                {/* </FormControl>
            <FormControl id="email"> */}
                <FormLabel>Email address</FormLabel>
                <Input
                  type="email"
                  id="email"
                  placeholder="name@email.com"
                  name="email"
                  value={state.email}
                  onChange={handleChange}
                />
                <FormHelperText>
                  I'll never share your email. I don't know how!
                </FormHelperText>
                {/* </FormControl>
            <FormControl id="password"> */}
                <FormLabel>Password</FormLabel>
                <Input
                  type="password"
                  placeholder="Choose a password"
                  name="password"
                  value={state.password}
                  onChange={handleChange}
                />
                <FormLabel>Photo</FormLabel>
                <Input
                  id="photo"
                  placeholder="Insert a link to your event photo"
                  value={state.photo}
                  name="photo"
                  onChange={handleChange}
                />
              </FormControl>
              <Stack spacing={10}>
                <FormControl id="type">
                  <FormLabel>Type of user</FormLabel>
                  <Select
                    name="host"
                    onChange={handleChange}
                    placeholder="Hosting or attending?"
                  >
                    <option>Guest</option>
                    <option>Host</option>
                  </Select>
                </FormControl>
                <Button
                  bg={'custom.200'}
                  color={'white'}
                  _hover={{
                    bg: 'custom.300',
                  }}
                  isDisabled={validateForm()}
                  onClick={handleSubmit}
                >
                  Sign up
                </Button>
              </Stack>
            </Stack>
          </Box>
        </motion.div>
      </Stack>
    </Flex>
  );
}
