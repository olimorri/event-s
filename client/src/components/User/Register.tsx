import React, { useState } from 'react';
import UsersApiService from '../../services/UsersApiService';
import auth from '../../utils/auth';
import {
  Flex,
  Box,
  FormControl,
  FormLabel,
  Input,
  Select,
  Stack,
  Button,
  Heading,
  FormHelperText,
} from '@chakra-ui/react';
import { motion } from 'framer-motion';
import { UserProps } from '../../interfaces/UserProps';
import { User } from '../../interfaces/User';
import { UserRequestError } from '../../interfaces/UserRequestError';

let email: string = '';
let password: string = '';
let firstName: string = '';
let lastName: string = '';
let photo: string = '';
let host: boolean = false;
let userTypeChosen: boolean = false;

const initialState = {
  email,
  password,
  firstName,
  lastName,
  photo,
  host,
};

export default function Register(props: UserProps) {
  const [state, setState] = useState(initialState);

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    let name: string = e.target.name;
    let value: string | boolean = e.target.value;

    setState((prevState) => ({
      ...prevState,
      [name]: value,
    }));
  };

  const handleUserSelection = (e: React.ChangeEvent<HTMLSelectElement>) => {
    userTypeChosen = true;
    let name: string = e.target.name;
    let value: string = e.target.value;

    if (name === 'host') {
      if (value !== 'Guest') {
        setState((prevState) => ({
          ...prevState,
          host: true,
        }));
      }
    }
  };

  const handleClick = async (e: React.MouseEvent) => {
    e.preventDefault();

    const { email, password, firstName, lastName, host, photo } = state;
    const user = { email, password, firstName, lastName, host, photo };
    const res = await UsersApiService.register(user);

    function isUser(res: any): res is User {
      return res.error === undefined;
    }

    function isUserRequest(res: any): res is UserRequestError {
      return res.error !== undefined;
    }

    if (isUser(res)) {
      props.setIsAuthenticated(true);
      auth.login(() => props.history.push('/profile'));
    } else if (isUserRequest(res)) {
      alert(`${res.message}`);
      setState(initialState);
    }
  };

  const validateForm = () => {
    if (
      !state.email ||
      !state.password ||
      !state.firstName ||
      !state.lastName ||
      userTypeChosen === false
    )
      return false;
  };
  return (
    <Flex
      minH={'100vh'}
      align={'center'}
      justify={'center'}
      bg={'gray.100'}
      bgImage="url('https://res.cloudinary.com/dujun1hoe/image/upload/v1615228154/event-s/gradient-background-26046-26731-hd-wallpapers.jpg_cenrqe.png')"
      bgSize="100%"
      backgroundRepeat="no-repeat"
    >
      <Stack spacing={8} w={'50%'} maxW={'lg'} py={12} px={6}>
        <motion.div
          initial={{
            opacity: 0,
            scale: 0,
          }}
          animate={{
            opacity: 1,
            scale: 1,
          }}
          transition={{
            // type: 'spring',
            stiffness: 100,
            // @ts-ignore
            ease: 'linear',
          }}
          exit={{
            opacity: 0,
            x: '-100vw',
          }}
        >
          <Stack align={'center'}>
            <Heading color={'white'} fontSize={'4xl'}>
              Create new account
            </Heading>
          </Stack>
          <Box rounded={'lg'} bg={'white'} boxShadow={'lg'} p={8}>
            <Stack spacing={4}>
              <FormControl id="first_name">
                <FormLabel>First name</FormLabel>
                <Input
                  id="first_name"
                  placeholder="First Name"
                  name="firstName"
                  value={state.firstName}
                  onChange={handleChange}
                />
                {/* </FormControl> */}
                {/* <FormControl id="last_name"> */}
                <FormLabel>Last name</FormLabel>
                <Input
                  id="last_name"
                  placeholder="Last Name"
                  name="lastName"
                  value={state.lastName}
                  onChange={handleChange}
                />
                {/* </FormControl>
            <FormControl id="email"> */}
                <FormLabel>Email address</FormLabel>
                <Input
                  type="email"
                  id="email"
                  placeholder="name@email.com"
                  name="email"
                  value={state.email}
                  onChange={handleChange}
                />
                <FormHelperText>
                  I'll never share your email. I don't know how!
                </FormHelperText>
                {/* </FormControl>
            <FormControl id="password"> */}
                <FormLabel>Password</FormLabel>
                <Input
                  type="password"
                  placeholder="Choose a password"
                  name="password"
                  value={state.password}
                  onChange={handleChange}
                />
                <FormLabel>Photo</FormLabel>
                <Input
                  id="photo"
                  placeholder="Insert a link to your event photo"
                  value={state.photo}
                  name="photo"
                  onChange={handleChange}
                />
              </FormControl>
              <Stack spacing={10}>
                <FormControl id="type">
                  <FormLabel>Type of user</FormLabel>
                  <Select
                    name="host"
                    onChange={handleUserSelection}
                    placeholder="Hosting or attending?"
                  >
                    <option>Guest</option>
                    <option>Host</option>
                  </Select>
                </FormControl>
                <Button
                  bg={'custom.200'}
                  color={'white'}
                  _hover={{
                    bg: 'custom.300',
                  }}
                  isDisabled={validateForm()}
                  onClick={handleClick}
                >
                  Sign up
                </Button>
              </Stack>
            </Stack>
          </Box>
        </motion.div>
      </Stack>
    </Flex>
  );
}
