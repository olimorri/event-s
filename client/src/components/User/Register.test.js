import { render, screen, cleanup, fireEvent } from '@testing-library/react';
import Register from './Register';

afterEach(cleanup);

describe('Snapshot', () => {
  it('should take a snapshot', () => {
    const { asFragment } = render(<Register />);

    expect(asFragment(<Register />)).toMatchSnapshot();
  });
});

describe('Unit Testing', () => {
  test('renders Register component', () => {
    render(<Register />);
    const registerTitle = screen.getByText(/create new account/i);
    expect(registerTitle).toBeInTheDocument();
  });

  test('renders first name input', () => {
    render(<Register />);
    const firstNameInput = screen.getByPlaceholderText(/First Name/i);
    expect(firstNameInput).toBeInTheDocument();
  });

  test('renders last name input', () => {
    render(<Register />);
    const lastNameInput = screen.getByPlaceholderText(/Last Name/i);
    expect(lastNameInput).toBeInTheDocument();
  });

  test('renders email input component', () => {
    render(<Register />);
    const emailInput = screen.getByPlaceholderText(/name@email.com/i);
    expect(emailInput).toBeInTheDocument();
  });

  test('renders password input component', () => {
    render(<Register />);
    const passwordInput = screen.getByPlaceholderText(/Choose a password/i);
    expect(passwordInput).toBeInTheDocument();
  });

  test('renders photo input component', () => {
    render(<Register />);
    const passwordInput = screen.getByLabelText(/Photo/i);
    expect(passwordInput).toBeInTheDocument();
  });

  test('renders type of user input component', () => {
    render(<Register />);
    const passwordInput = screen.getByLabelText(/Hosting or Attending?/i);
    expect(passwordInput).toBeInTheDocument();
  });

  test('renders user input options', () => {
    render(<Register />);
    const guestOption = screen.getByRole('option', { name: /^Guest$/i });
    const hostOption = screen.getByRole('option', { name: /^Host$/i });
    expect(guestOption && hostOption).toBeInTheDocument();
  });

  test('renders sign up button component', () => {
    render(<Register />);
    const signUp = screen.getByRole('button', { name: /sign up/i });
    expect(signUp).toBeInTheDocument();
  });
});

// https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques#roles
