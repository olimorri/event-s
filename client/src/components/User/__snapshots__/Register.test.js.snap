// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot should take a snapshot 1`] = `
<DocumentFragment>
  <div
    class="css-1rpqb1v"
  >
    <div
      class="chakra-stack css-rfhkiz"
    >
      <div
        style="opacity: 0; transform: scale(0) translateZ(0);"
      >
        <div
          class="chakra-stack css-1xbaot1"
        >
          <h2
            class="chakra-heading css-tyi3eh"
          >
            Create new account
          </h2>
        </div>
        <div
          class="css-3hziw7"
        >
          <div
            class="chakra-stack css-ppbuf4"
          >
            <div
              class="chakra-form-control css-1kxonj9"
              role="group"
            >
              <label
                class="chakra-form__label css-uvho5g"
                for="first_name"
                id="first_name-label"
              >
                First name
              </label>
              <input
                aria-describedby="first_name-helptext"
                class="chakra-input css-0"
                id="first_name"
                name="firstName"
                placeholder="First Name"
                value=""
              />
              <label
                class="chakra-form__label css-uvho5g"
                for="first_name"
                id="first_name-label"
              >
                Last name
              </label>
              <input
                aria-describedby="first_name-helptext"
                class="chakra-input css-0"
                id="last_name"
                name="lastName"
                placeholder="Last Name"
                value=""
              />
              <label
                class="chakra-form__label css-uvho5g"
                for="first_name"
                id="first_name-label"
              >
                Email address
              </label>
              <input
                aria-describedby="first_name-helptext"
                class="chakra-input css-0"
                id="email"
                name="email"
                placeholder="name@email.com"
                type="email"
                value=""
              />
              <div
                class="chakra-form__helper-text css-0"
                id="first_name-helptext"
              >
                I'll never share your email. I don't know how!
              </div>
              <label
                class="chakra-form__label css-uvho5g"
                for="first_name"
                id="first_name-label"
              >
                Password
              </label>
              <input
                aria-describedby="first_name-helptext"
                class="chakra-input css-0"
                id="first_name"
                name="password"
                placeholder="Choose a password"
                type="password"
                value=""
              />
              <label
                class="chakra-form__label css-uvho5g"
                for="first_name"
                id="first_name-label"
              >
                Photo
              </label>
              <input
                aria-describedby="first_name-helptext"
                class="chakra-input css-0"
                id="photo"
                name="photo"
                placeholder="Insert a link to your event photo"
                value=""
              />
            </div>
            <div
              class="chakra-stack css-1re04b2"
            >
              <div
                class="chakra-form-control css-1kxonj9"
                role="group"
              >
                <label
                  class="chakra-form__label css-uvho5g"
                  for="type"
                  id="type-label"
                >
                  Hosting or Attending?
                </label>
                <div
                  class="chakra-select__wrapper css-42b2qy"
                >
                  <select
                    class="chakra-select css-1p4gbel"
                    id="type"
                    name="host"
                  >
                    <option
                      selected=""
                    >
                      Guest
                    </option>
                    <option>
                      Host
                    </option>
                  </select>
                  <div
                    class="chakra-select__icon-wrapper css-162mkon"
                  >
                    <svg
                      aria-hidden="true"
                      class="chakra-select__icon"
                      focusable="false"
                      role="presentation"
                      style="width: 1em; height: 1em; color: currentColor;"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                        fill="currentColor"
                      />
                    </svg>
                  </div>
                </div>
              </div>
              <button
                class="chakra-button css-1fvu2nh"
                type="button"
              >
                Sign up
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
